internal/validate/struct.go

```diff
@@ -0,0 +1,146 @@
+package validate
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/getkin/kin-openapi/openapi3"
+)
+
+// Validate that the values in the struct are valid according to the validation rules.
+// If validation fails the error will be of type Error.
+func Validate(req Request) error {
+	value := reflect.ValueOf(req)
+	if value.Kind() != reflect.Ptr || value.Elem().Kind() != reflect.Struct {
+		return fmt.Errorf("request argument must be pointer to struct, not %T", req)
+	}
+	value = value.Elem()
+
+	fieldErrors := make(map[string][]string)
+
+	for _, rule := range req.ValidationRules() {
+		fv := rule.fieldValue()
+		if fv.Kind() != reflect.Ptr {
+			return fmt.Errorf("field value must be a pointer, not %v", fv.Kind())
+		}
+		ft := findStructField(value, fv)
+		if ft == nil {
+			return fmt.Errorf("could not locate field for %v", fv.Type())
+		}
+		jsonName := jsonNameFromStructField(ft)
```

Comment:
I think this is the bit I'm least thrilled about. This approach was inspired by `ozzo-validate`, but it only works when the `ValidationRules` method has a pointer receiver.

I'm thinking it would be easier to have the validation rules accept the name of the json field instead of trying to look up the name by matching pointer addresses.

If we take that approach we can delete `findStructField`, `jsonNameFromStructField`, `StructRule`, and simplify this function and `RulesToMap`. The other rules would take an extra argument of `jsonName string`, and the field value would not need to be a pointer.